sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java

```diff
@@ -0,0 +1,500 @@
+      Document stats = mongoDatabase.runCommand(stat);
+      estimatedSizeBytes = Long.valueOf(stats.get("size").toString());
+      long avgSize = Long.valueOf(stats.get("avgObjSize").toString());
+      if (filter != null && !filter.isEmpty()) {
+        // as the user is using a filter, we improve the estimated size
+        // by counting number of document returned by the filter
+        // and using the average object size from collStats
+        try {
+          estimatedSizeBytes = getFilterResultByteSize(mongoCollection, avgSize);
+        } catch (Exception e) {
+          LOGGER.warn("Can't estimate size", e);
+        }
+      }
+      return estimatedSizeBytes;
+    }
+
+    private long getFilterResultByteSize(MongoCollection mongoCollection, long avgSize)
+        throws Exception {
+      // count the number of documents selected by the filter and estimate the size of the
+      // filter using the average object size * the number of documents
+      long totalFilterByteSize = 0L;
+      Document bson = Document.parse(filter);
+      long countFilter = mongoCollection.count(bson);
+      totalFilterByteSize = countFilter * avgSize;
+      return totalFilterByteSize;
+    }
+
+    @Override
+    public List<BoundedSource> splitIntoBundles(long desiredBundleSizeBytes,
+                                                PipelineOptions options) {
+      MongoClient mongoClient = new MongoClient();
+      MongoDatabase mongoDatabase = mongoClient.getDatabase(database);
+
+      ArrayList splitKeys = null;
+      if (numSplits > 0) {
+        // the user defines his desired number of splits
+        // calculate the bundle size
+        long estimatedSizeBytes = getEstimatedSizeBytes(options);
+        desiredBundleSizeBytes = estimatedSizeBytes / numSplits;
+      }
+
+      // now we have the desized bundle size (provided by user or provided by the runner)
+      // we use Mongo splitVector command to get the split keys
+      BasicDBObject splitVectorCommand = new BasicDBObject();
+      splitVectorCommand.append("splitVector", database + "." + collection);
+      splitVectorCommand.append("keyPattern", new BasicDBObject().append("_id", 1));
+      splitVectorCommand.append("force", false);
+      // maxChunkSize is the Mongo partition size in MB
+      splitVectorCommand.append("maxChunkSize", desiredBundleSizeBytes / 1024 / 1024);
+      Document splitVectorCommandResult = mongoDatabase.runCommand(splitVectorCommand);
+      splitKeys = (ArrayList) splitVectorCommandResult.get("splitKeys");
+
+      return createSourceList(splitKeys);
+    }
+
+    /**
+     * Create a sources list. Each source in the list has a subset of keys to deal with (split).
+     *
+     * @param splitKeys List of split keys to create sources.
+     * @return The list of sources.
+     */
+    private List<BoundedSource> createSourceList(ArrayList splitKeys) {
+      List<BoundedSource> splitSourceList = new ArrayList<>();
+      if (splitKeys == null) {
```

Comment:
Better make this method take a guaranteed non-null argument, and do this null checking at the call site. This will make the invariants of the current method stricter and hence easier to understand, and will also reduce nesting in code.